add 'list' command to query package-version in snapshot

juhp opened the pull request:
currently not working for core ghc libs with resolver

Please include the following checklist in your PR:

* [x] Any changes that could be relevant to users have been recorded in the ChangeLog.md
* [x] The documentation has been updated, if necessary.

Please also shortly describe how you tested your change. Bonus points for added tests!

```
$ stack list zlib
zlib-0.6.2.2
$ stack --resolver lts-13 list zlib
Selected resolver: lts-13.30
zlib-0.6.2
```

juhp commented:
I would appreciate feedback, comments, code review of this.

qrilka reviewed (approved):
LGTM but I guess we need to have an opinion from one of Stack's maintainers as well to merge this in. Also a changelog entry looks to be needed in that case and probably mentioning it in the docs.

qrilka commented:
Also style job error needs to be fixed

qrilka commented:
@juhp I've mentioned this PR in our discussion with @snoyberg and it looks like there are no particular objections to merge this. Could you address the minor issues about it I pointed above?

juhp commented:
Sure, thanks, @qrilka, I will try to get to it soon.

(Maybe I will try to check on core library handling too at the same time, though I think it's already quite useful.)

juhp commented:
I add a Changelog entry and short documentation. Also fixed the hlint warning, and updated package.yaml.
Please have a look, this should be ready now.

juhp commented:
I slightly tweaked the documentation and changelog.

qrilka commented on src/Stack/List.hs:
```diff
@@ -0,0 +1,77 @@
+    toLoc | Just snapshot <- mSnapshot = toLocSnapshot snapshot
+          | otherwise = toLocNoSnapshot
+
+    -- FIXME fallback to global or project resolver
```
@juhp what do you mean by this FIXME? Do you want to change the way `stack list` works with no resolver specified? Do you think the way it works in this implementation is bad in some sense?

juhp replied:
Good catch - I was looking at this too earlier, and then forgot to remove it.
For now I feel falling back to the Hackage is probably good enough and useful , since `stack list` is not a project command.
So I think I will just drop the comment (I had left it there as a placeholder/reminder about this decision).

qrilka reviewed (commented):
I'm OK with merging this but would like to understand the point of FIXME in the code.
Thank you.

juhp commented:
Okay, I dropped the comment, thanks

qrilka commented:
Great, thanks

juhp commented:
Thanks @qrilka for reviewing and helping to get this ready

mpilgrem commented:
@juhp, may I clarify how this command is intended to work when no packages are specified at the command line? If I command:

~~~text
stack --resolver nightly-2022-09-23 list
~~~

it yields a long list of what looks like the packages in `nightly-2022-09-23`, save that it excludes the packages that come with GHC on Unix-like operating systems (`base`, `Cabal`, `time` etc) - `Win32-2.12.0.1` is listed.

If I command:

~~~text
stack list
~~~

it yields nothing. I was rather hoping it would yield the snapshot specified by Stack's project-level configuration file (`stack.yaml`). Is nothing the intent?

juhp commented:
I believe that is the current behaviour yes. I mean I would like the boot libraries to be listed too, but I didn't manage to wrangle that - if it's straightforward to do that would be great, thanks!

mpilgrem commented:
@juhp, many thanks. For now. I've updated the online documentation to describe more fully the existing command.
